File: backend/src/schemas/project.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
import uuid
from pydantic import field_serializer


class ProjectBase(BaseModel):
    name: str
    slug: str
    description: Optional[str] = None
    due_date: Optional[datetime] = None
    priority: Optional[str] = "medium"  # low, medium, high
    completed: bool = False


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(ProjectBase):
    name: Optional[str] = None
    completed: Optional[bool] = None


class ProjectResponse(ProjectBase):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

    @field_serializer('id', 'user_id')
    def serialize_uuid(self, value):
        if value is not None:
            return str(value)
        return value

File: frontend/lib/types/project.ts
// frontend/lib/types/project.ts
export interface Project {
  id: string;
  name: string;
  slug: string;
  description?: string;
  user_id: string;
  due_date?: string; // ISO date string
  priority?: 'low' | 'medium' | 'high';
  completed: boolean;
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  // Analytics/Computed fields
  members?: number;
  progress?: number;
  tasks_completed?: number;
  tasks_total?: number;
}

export interface ProjectCreate {
  name: string;
  slug: string;
  description?: string;
  due_date?: string;
  priority?: 'low' | 'medium' | 'high';
  completed?: boolean;
}

export interface ProjectUpdate {
  name?: string;
  slug?: string;
  description?: string;
  due_date?: string;
  priority?: 'low' | 'medium' | 'high';
  completed?: boolean;
}

export interface ProjectToggleComplete {
  completed: boolean;
}

export interface ProjectListResponse {
  projects: Project[];
  total: number;
  offset: number;
  limit: number;
}

export interface ProjectAPIResponse {
  id: string;
  name: string;
  slug: string;
  description?: string;
  user_id: string;
  due_date?: string;
  priority?: 'low' | 'medium' | 'high';
  completed: boolean;
  created_at: string;
  updated_at: string;
}